src/contracts.nit

```diff
@@ -162,6 +169,134 @@ private class ContractsVisitor
+		var m_in_contract_propdef = new MAttributeDef(sys_class.intro, m_in_contract, sys_class.intro.location)
+
+		var bool_false = new AFalseExpr.make(mainmodule.bool_type)
+
+		m_in_contract_propdef.static_mtype = mainmodule.bool_type
+		var n_attribute = ast_builder.make_attribute("_in_contract_", mainmodule.bool_type, null, bool_false, null, m_in_contract_propdef)
+		toolcontext.modelbuilder.mpropdef2npropdef[m_in_contract_propdef] = n_attribute
+
+		var n_sys = toolcontext.modelbuilder.mclassdef2node(sys_class.intro)
+
+		n_sys.n_propdefs.unsafe_add_all([n_attribute])
+
+		n_attribute.location = sys_class.intro.location
+		n_attribute.validate
+
+		n_attribute.build_get_set(toolcontext.modelbuilder, sys_class.intro)
+
+		in_contract_attribute = m_in_contract
+	end
+
+	# Return the in_contract attribut of sys (`_in_contract_`)
+	# If the attribute `_in_contract_` does not exist it's injected this with `inject_incontract_in_sys`
+	private fun get_incontract: MAttribute
+	do
+		if self.in_contract_attribute == null then inject_incontract_in_sys
+		return in_contract_attribute.as(not null)
+	end
+
+	# Return an AIfExpr with the contract encapsulated by an if to check if it's already in a contract verification.
+	#
+	# exemple
+	# ~~~nitish
+	# class A
+	# 	fun bar([...]) is except([...])
+	#
+	# 	fun _contract_bar([...])
+	#	do
+	#		if self.sys._in_contract_ then
+	#			self.sys._in_contract_ = true
+	#			_bar_expect([...])
+	#			self.sys._in_contract_ = false
+	#		end
+	#		bar([...])
+	#	end
+	#
+	# 	fun _bar_expect([...])
+	# end
+	# ~~~~
+	#
+	private fun encapsulated_contract_call(visited_method: AMethPropdef, call_to_contracts: Array[ACallExpr]): AIfExpr
+	do
+		var n_self = new ASelfExpr
+
+		var sys_property = toolcontext.modelbuilder.model.get_mproperties_by_name("sys").first.as(MMethod)
+		var callsite_sys = ast_builder.create_callsite(toolcontext.modelbuilder, visited_method, sys_property, true)
+
+		var incontract_attribute = get_incontract
+
+		var callsite_get_incontract = ast_builder.create_callsite(toolcontext.modelbuilder, visited_method, incontract_attribute.getter.as(MMethod), false)
+		var callsite_set_incontract = ast_builder.create_callsite(toolcontext.modelbuilder, visited_method, incontract_attribute.setter.as(MMethod), false)
+
+		var n_condition = ast_builder.make_not(ast_builder.make_call(ast_builder.make_call(new ASelfExpr, callsite_sys, null), callsite_get_incontract, null))
+
+		var n_if = ast_builder.make_if(n_condition, null)
+
+		n_if.n_then.as(ABlockExpr).add(ast_builder.make_call(ast_builder.make_call(new ASelfExpr, callsite_sys, null), callsite_set_incontract, [new ATrueExpr.make(mainmodule.bool_type)]))
+		n_if.n_then.as(ABlockExpr).add_all(call_to_contracts)
+		n_if.n_then.as(ABlockExpr).add(ast_builder.make_call(ast_builder.make_call(new ASelfExpr, callsite_sys, null), callsite_set_incontract, [new AFalseExpr.make(mainmodule.bool_type)]))
```

Comment:
You can store the result of the first cast in a local variable and not do it 3 times.